### PR TITLE
Add layers to matrices and implement `apply` function

### LINKED ISSUE #3: Add the ability for a model to pass data through layers of matrices

### LINKED ISSUE #4: Add bias vectors and activation functions
This is a minor one, to be completed after #3 is merged.

### PR DESCRIPTION
**Pull request by adamhutchings**

### CODE REVIEW

File: core/include/jml/model.hpp [hunk 1]
```diff
@@ -10,10 +10,38 @@
...
+  std:unique_ptr<Vector> apply(Vector& in);
 
+private:
+  Model_Layer* layers_arr; // Array of layer addresses
+  unsigned int num_layers; // A.k.a. the number of entries in layers_arr
 };
 
 }
+
+std:unique_ptr<Vector> Model::apply(Vector& in) {
```
Comment (adamhutchings) [inline core/include/jml/model.hpp:139]: This should go in a `.cpp` file, not a `.hpp` file. There should be one already made.

File: core/include/jml/model.hpp [hunk 2]
```diff
@@ -10,10 +10,38 @@
...
 
+class Model_Layer {
+  friend class Layer_Actv_Funcs_List;
+
+private:
+  Matrix* layer_matrix;
+  Vector* layer_bias_vector;
+  // Array of pointers to function pointers (because lonely function pointers might not be contiguous in memory)
+  // The size of this will be dictated by the associated matrix
+  double (*layer_actv_funcs_arr[])();
```
Comment (adamhutchings) [inline core/include/jml/model.hpp]: What do we need an entire array of function pointers for? Each layer should only have one activation function.

File: core/include/jml/model.hpp [hunk 3]
```diff
@@ -10,10 +10,38 @@
...
+};
+
+// Transition class for loading in/saving model, includes parameters and activation func ID's (enum <-> func pointer mappings?)
+class Model_Layer_Compressed {
+  friend class Model_Layer;
+
+};
+
 class Model {
+  friend class Model_Layer;
```
Comment (adamhutchings) [inline core/include/jml/model.hpp]: Also, when the model is initialized, we should actually create the model layers. (And when the model is destroyed in `~Model`, it should free the model layers as well. This means passing in more data to the constructor, which we can tidy up in the future.

File: core/include/jml/model.hpp (hunk 2 above)
Comment (mawroblewski1) [reply core/include/jml/model.hpp]: For mixing/matching, and directly including function pointers might mean more memory space but faster accesses than enum -> function pointer lookups

Comment (adamhutchings) [reply core/include/jml/model.hpp]: Explain what you mean by mixing/matching.

Comment (mawroblewski1) [reply core/include/jml/model.hpp]: Each layer has multiple perceptrons - with mixing and matching, the activation functions within each layer need not be identical which could produce results of experimental interest

https://discord.com/channels/909325954638299136/1153015102568677417/1156740693407039528

https://discord.com/channels/909325954638299136/1153015102568677417/1157907596745904199

Comment (adamhutchings) [reply core/include/jml/model.hpp]: It is possible that we could keep it as one activation function for now and possibly experiment with that later / for other layer types?

Comment (mawroblewski1) [reply core/include/jml/model.hpp]: Jake's abstract class ActivationFunction in activation_functions.hpp will permit a very easy, modular implementation of this

Comment (adamhutchings) [reply core/include/jml/model.hpp]: Okay, then keeping it sounds good. But do add the implementations of the class itself and nt just the header, and we should be good to go.

**Review by adamhutchings — COMMENTED** [review]
(This is meant to say "changes requested" but I can't do that on my own PR.)

This is all good, except that we need an implementation for the `apply` method, and added handling of the layers inside the model setup/teardown methods.

**Review by JakeRoggenbuck — CHANGES_REQUESTED** [review]
See https://github.com/jabacat/jml/pull/12#pullrequestreview-1651932862

**Comment by adamhutchings** [comment]
Closes #3 and #4.

**Comment by Sophon96** [comment]
> Closes #3 and #4.

This only seems to work in the original PR post ;) I've added them manually (you can do it too on the right toolbar)

**Comment by adamhutchings** [comment]
I will not push to Marcin's fork, and so I will close this PR and re-open it from my own branch.

**Pull request closed without merging**